src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -241,19 +253,215 @@ uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md,
+
+    iter = kh_get(rkeys, &md->lru.rkeys_hash, head->rkey);
+    ucs_assert(iter != kh_end(&md->lru.rkeys_hash));
+    kh_del(rkeys, &md->lru.rkeys_hash, iter);
+    ucs_free(head);
+}
+
+static ucs_status_t
+uct_ib_md_mlx5_devx_mr_lru_push(uct_ib_mlx5_md_t *md, uint32_t rkey)
+{
+    static size_t counter              = 0;
+    uct_ib_mlx5_mem_lru_entry_t *entry = ucs_malloc(sizeof(*entry),
+                                                    "rkey_lru_entry");
+    ucs_kh_put_t res;
+
+    if (entry == NULL) {
+        ucs_error("Cannot allocate rkey LRU entry");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    entry->rkey = rkey;
+    ucs_list_add_tail(&md->lru.rkeys_list, &entry->link);
+    kh_put(rkeys, &md->lru.rkeys_hash, rkey, &res);
+    ucs_assert((res == UCS_KH_PUT_BUCKET_EMPTY) ||
+               (res == UCS_KH_PUT_BUCKET_CLEAR));
+    if (kh_size(&md->lru.rkeys_hash) > md->super.config.max_mr_gc_queue) {
+        uct_ib_mlx5_devx_md_mr_lru_pop(md);
+    }
+
+    if ((++counter % (md->super.config.max_mr_gc_queue)) == 0) {
+        md->mkey_tag += 1;
+        md->mkey_tag %= 251;
+    }
+
+    ucs_trace("mr lru size=%d push: %x", kh_size(&md->lru.rkeys_hash), rkey);
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_ib_md_mlx5_devx_mr_lru_test(uct_ib_mlx5_md_t *md, uint32_t rkey)
+{
+    return (kh_get(rkeys, &md->lru.rkeys_hash, rkey) ==
+            kh_end(&md->lru.rkeys_hash)) ? UCS_OK : UCS_ERR_ALREADY_EXISTS;
+}
+
+static void uct_ib_mlx5_devx_mr_lru_init(uct_ib_mlx5_md_t *md)
+{
+    ucs_list_head_init(&md->lru.rkeys_list);
+    kh_init_inplace(rkeys, &md->lru.rkeys_hash);
+}
+
+static void uct_ib_mlx5_devx_mr_lru_cleanup(uct_ib_mlx5_md_t *md)
+{
+    while (kh_size(&md->lru.rkeys_hash) > 0) {
+        uct_ib_mlx5_devx_md_mr_lru_pop(md);
+    }
+
+    kh_destroy_inplace(rkeys, &md->lru.rkeys_hash);
+}
+
+static ucs_status_t uct_ib_mlx5_devx_reg_indirect_key(uct_ib_md_t *ibmd,
+                                                      uct_ib_mem_t *ib_memh)
+{
+    uct_ib_mlx5_md_t *md    = ucs_derived_of(ibmd, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    ucs_queue_head_t tmp_mrs;
+    uct_ib_mlx5_mem_tmp_entry_t *tmp_mr;
+    ucs_status_t status;
+    int ret;
+
+    if (!(md->flags & UCT_IB_MLX5_MD_FLAG_KSM)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    ucs_assert(uct_ib_memh_get_atomic_base_mr_type(ib_memh) ==
+               UCT_IB_MR_DEFAULT);
+    ucs_queue_head_init(&tmp_mrs);
+
+    for (;;) {
+        status = uct_ib_mlx5_devx_reg_ksm_data_contig(md,
+                                                      &memh->mrs[UCT_IB_MR_DEFAULT],
+                                                      0, &memh->indirect_dvmr,
+                                                      &memh->super.indirect_rkey);
+        if (status != UCS_OK) {
+            break;
+        }
+
+        status = uct_ib_md_mlx5_devx_mr_lru_test(md, memh->super.indirect_rkey);
```

Comment:
maybe unite test and uct_ib_md_mlx5_devx_mr_lru_push to single function: it accepts key number, key object (can be NULL), and returns status if it already existed